File: 풀이/Juhee/done/BOJ/[DP]동전1.js
// https://www.acmicpc.net/problem/2293 동전1

let fs = require("fs");
let input = fs
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "test.txt")
  .toString()
  .trim()
  .split("\n");

const solution = (input) => {
  const [N, K] = input.shift().split(" ").map(Number); // 동전 가지수 n, 목표 k원
  const coins = input.map(Number);

  const dp = Array.from({ length: K + 1 }, () => 0);
  dp[0] = 1;

  // dp[i] = i를 만들 수 있는 경우의 수
  for (const coin of coins) {
    for (let now = coin; now <= K; now++) {
      dp[now] += dp[now - coin];
    }
  }

  return dp[K];
};

console.log(solution(input));

File: 풀이/Jihoon/done/[DP]동전1.py
# 본인이 선택한 언어로 풀이해주세요.
# https://www.acmicpc.net/problem/2293
n,k = map(int, input().split())
coins = []
for i in range(n):
    coins.append(int(input()))

coins.sort() # 오름차순 정렬

dp = [0 for i in range(k+1)]
dp[0] = 1   # 동전을 1개만 쓴 경우

for coin in coins:
    for j in range(coin, k+1):
        if j-coin >= 0:
            dp[j] = dp[j] + dp[j-coin]
            print(dp)
    print("\n")

print(dp[k])